WalletWasabi/WabiSabi/Client/OwnershipProofValidator.cs

```diff
@@ -0,0 +1,113 @@
+using NBitcoin;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+using WalletWasabi.Blockchain.Transactions;
+using WalletWasabi.Crypto;
+using WalletWasabi.Stores;
+using WalletWasabi.Wallets;
+
+namespace WalletWasabi.WabiSabi.Client
+{
+	// OwnershipProofValidator validates the ownership proofs provided by the coordinator.
+	public class OwnershipProofValidator
+	{
+		public OwnershipProofValidator(IndexStore indexStore, TransactionStore transactionStore, IBlockProvider blockProvider)
+		{
+			IndexStore = indexStore;
+			TransactionStore = transactionStore;
+			BlockProvider = blockProvider;
+		}
+
+		public IndexStore IndexStore { get; }
+		public TransactionStore TransactionStore { get; }
+		public IBlockProvider BlockProvider { get; }
+
+		// Verifies the coins and ownership proofs. These proofs are provided by the coordinator who should
+		// have validate them before which means that when dealing with a honest coordinator all these proof
+		// are valid.
+		//
+		// In case one proof is invalid we have evidence enough to know the coordinator is malicious and we
+		// have to abort the process, notify the user and ban the coordinator.
+		public async ValueTask<int> VerifyOtherAlicesOwnershipProofsAsync(
+			uint256 roundId,
+			IEnumerable<(Coin Coin, OwnershipProof OwnershipProof)> othersCoins,
+			int stopAfter,
+			CancellationToken cancellationToken)
+		{
+			var proofChannel = Channel.CreateBounded<(Coin, Coin?, OwnershipProof)>(10);
```

Comment:
What happens if we write more than 10?

Reply:
Do we need to use Channel.CreateBounded? A simple list is not enough?

Reply:
writing more than the channel bound will block the writer

Reply:
> Do we need to use Channel.CreateBounded? A simple list is not enough?

Good question. All channels should be bounded by default while unbounded channels should be the exception. Here having and unbounded channel wouldn't be problematic at all because the number of elements that can be pushed is small and because the consumer is much faster than the producer, however that's not true by design but by accident because the blocks are downloaded in a serial way.

Reply:
FWIW I like Go's default behavior which is to default to a bound of 0. Is that supported in C#? i can't tell based on the docs. This makes means backpressure is immediate, and the chance of having harder to catch deadlocks due to writes blocking less often is also reduced.

Also note that the bound is on how many non-blocking writes will succeed, but blocking writes will still consume more memory (with an implicit bound being the total number of active threads) unless `FullMode` is set to one of the dropping behaviors, only then is the space complexity truly bounded by the given value.

Reply:
IDK. I would use bounded to 1.